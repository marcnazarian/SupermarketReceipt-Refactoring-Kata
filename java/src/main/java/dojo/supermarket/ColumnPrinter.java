package dojo.supermarket;

public class ColumnPrinter {
    private int columns;

    public ColumnPrinter(int columns) {

        this.columns = columns;
    }

    String getWhitespace(int whitespaceSize) {
        StringBuilder whitespace = new StringBuilder();
        for (int i = 0; i < whitespaceSize; i++) {
            whitespace.append(" ");
        }
        return whitespace.toString();
    }

    String formatColumns(String leftColumn, String rightColumn) {
        String whitespaces = getWhitespace(this.columns - leftColumn.length() - rightColumn.length());
        return leftColumn + whitespaces + rightColumn;
    }
}
