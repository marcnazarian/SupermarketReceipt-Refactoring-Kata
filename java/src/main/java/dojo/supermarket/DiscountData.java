package dojo.supermarket;

import dojo.supermarket.model.Discount;

public class DiscountData {
    private Discount discount;
    private String productPresentation;
    private String description;
    private double discountAmount;

    public DiscountData(String productPresentation, String description, double discountAmount) {
        this.productPresentation = productPresentation;
        this.description = description;
        this.discountAmount = discountAmount;
    }

    public String getProductPresentation() {
        return productPresentation;
    }

    public String getDescription() {
        return description;
    }

    public double getDiscountAmount() {
        return discountAmount;
    }

}
