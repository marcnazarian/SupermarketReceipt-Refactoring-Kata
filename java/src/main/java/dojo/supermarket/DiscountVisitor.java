package dojo.supermarket;

import dojo.supermarket.model.Product;

public class DiscountVisitor {
    public void visitProductname(Product product) {

    }

    public void visitDescription(String description) {

    }

    public void visitAmount(double discountAmount) {

    }
}
