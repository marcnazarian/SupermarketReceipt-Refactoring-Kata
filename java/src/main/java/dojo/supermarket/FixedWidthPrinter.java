package dojo.supermarket;

public class FixedWidthPrinter {
    private int columns;

    public FixedWidthPrinter(int columns) {

        this.columns = columns;
    }

    private static String getWhitespace(int whitespaceSize) {
        StringBuilder whitespace = new StringBuilder();
        for (int i = 0; i < whitespaceSize; i++) {
            whitespace.append(" ");
        }
        return whitespace.toString();
    }

    public String formatColumns(String leftColumn, String rightColumn) {
        String whitespaces = getWhitespace(this.columns - leftColumn.length() - rightColumn.length());
        return leftColumn +
                whitespaces + rightColumn;
    }
}
