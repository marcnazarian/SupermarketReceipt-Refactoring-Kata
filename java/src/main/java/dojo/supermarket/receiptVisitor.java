package dojo.supermarket;

public class receiptVisitor {
    private int columns;
    private StringBuilder receiptText = new StringBuilder();

    public receiptVisitor(int columns) {

        this.columns = columns;
    }


    public void addItemQuantity(String line) {
        receiptText.append(line);

    }
    public void addReceiptItem(String name, String price) {
        String line = formatColumns(name, price);
        receiptText.append( line + "\n");

    }


    public void addDiscount(String description, String pricePresentation) {
        String line = formatColumns(description, pricePresentation);
        receiptText.append( line + "\n");

    }

    public void appendTotalSection(String total, String pricePresentation) {
        String line = formatColumns(total, pricePresentation);
        receiptText.append("\n");
        receiptText.append(line);
    }

    public String build() {
        return receiptText.toString();
    }


    private String formatColumns(String leftColumn, String rightColumn) {
        String whitespaces = getWhitespace(this.columns - leftColumn.length() - rightColumn.length());
        return leftColumn +
                whitespaces + rightColumn;
    }

    private static String getWhitespace(int whitespaceSize) {
        StringBuilder whitespace = new StringBuilder();
        for (int i = 0; i < whitespaceSize; i++) {
            whitespace.append(" ");
        }
        return whitespace.toString();
    }

}
