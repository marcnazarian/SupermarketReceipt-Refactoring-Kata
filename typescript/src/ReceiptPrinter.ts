import {ProductUnit} from "./model/ProductUnit"
import {ReceiptItem} from "./model/ReceiptItem"
import {Receipt} from "./model/Receipt"

export class ReceiptPrinter {

    public constructor(private readonly columns: number = 40) {
    }

    public printReceipt( receipt: Receipt): string {
        const result = "";
        for (const item of receipt.getItems()) {
            let price = new Intl.NumberFormat('en-UK', {minimumFractionDigits: 2}).format(item.getTotalPrice());
            let quantity = ReceiptPrinter.presentQuantity(item);
            let name = item.getProduct().getName();
            let unitPrice = new Intl.NumberFormat('en-UK', {minimumFractionDigits: 2}).format(item.getPrice());

            let whitespaceSize = this.columns - name.length - price.length;
            let line = name + ReceiptPrinter.getWhitespace(whitespaceSize) + price + "\n";

            if (item.getQuantity() != 1) {
                line += "  " + unitPrice + " * " + quantity + "\n";
            }
            result.concat(line);
        }
        for (const discount of receipt.getDiscounts()) {
            let productPresentation = discount.getProduct().getName();
            let pricePresentation = new Intl.NumberFormat('en-UK', {minimumFractionDigits: 2}).format(discount.getDiscountAmount());
            let description = discount.getDescription();
            result.concat(description);
            result.concat("(");
            result.concat(productPresentation);
            result.concat(")");
            result.concat(ReceiptPrinter.getWhitespace(this.columns - 3 - productPresentation.length - description.length - pricePresentation.length));
            result.concat("-");
            result.concat(pricePresentation);
            result.concat("\n");
        }
        result.concat("\n");
        let pricePresentation = new Intl.NumberFormat('en-UK', {minimumFractionDigits: 2}).format(receipt.getTotalPrice());
        let total = "Total: ";
        let whitespace = ReceiptPrinter.getWhitespace(this.columns - total.length - pricePresentation.length);
        result.concat(total).concat(whitespace).concat(pricePresentation);
        return result;
    }

    private static presentQuantity( item: ReceiptItem): string  {
        return ProductUnit.Each == item.getProduct().getUnit()
            // TODO make sure this is the simplest way to make something similar to the java version
                ? new Intl.NumberFormat('en-UK', {maximumFractionDigits: 0}).format(item.getQuantity())
                : new Intl.NumberFormat('en-UK', {minimumFractionDigits: 3}).format(item.getQuantity());
    }

    private static getWhitespace(whitespaceSize: number): string {
        const whitespace = "";
        for (let i = 0; i < whitespaceSize; i++) {
            whitespace.concat(" ");
        }
        return whitespace;
    }
}
