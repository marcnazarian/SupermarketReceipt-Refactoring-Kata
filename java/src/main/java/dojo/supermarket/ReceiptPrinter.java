package dojo.supermarket;

import dojo.supermarket.model.*;

public class ReceiptPrinter {

    private final int columns;

    public ReceiptPrinter(int columns) {
        this.columns = columns;
    }

    public String printReceipt(Receipt receipt) {
        StringBuilder result = new StringBuilder();
        for (ReceiptItem item : receipt.getItems()) {
            String line = receiptItemLine(item);
            result.append(line);
        }
        for (Discount discount : receipt.getDiscounts()) {
            String discountline = discountLine(discount);
            result.append(discountline + "\n");
        }
        result.append("\n");
        String line = receiptTotalSection(receipt);
        result.append(line);
        return result.toString();
    }

    private String discountLine(Discount discount) {
        DiscountVisitor discountVisitor = new DiscountVisitor();
        discountVisitor.visitProductname(discount.getProduct());
        discountVisitor.visitDescription(discount.getDescription());
        discountVisitor.visitAmount(discount.getDiscountAmount());

        String productPresentation = discount.getProduct().getName();
        String pricePresentation = String.format("%.2f", discount.getDiscountAmount());
        String description = discount.getDescription();
        String leftColumn = description + "(" + productPresentation + ")";
        String rightColumn = "-" + pricePresentation;
        return formatColumns(leftColumn, rightColumn);
    }

    private String formatColumns(String leftColumn, String rightColumn) {
        String whitespaces = getWhitespace(this.columns - leftColumn.length() - rightColumn.length());
        return leftColumn +
                whitespaces + rightColumn;
    }

    private String receiptTotalSection(Receipt receipt) {
        String pricePresentation = String.format("%.2f", (double) receipt.getTotalPrice());
        String total = "Total: ";
        return formatColumns(total, pricePresentation);
    }

    private String receiptItemLine(ReceiptItem item) {

        String price = String.format("%.2f", item.getTotalPrice());
        String name = item.getProduct().getName();

        String line1 = formatColumns(name, price);

        String line = line1 + "\n";

        if (item.getQuantity() != 1) {
            String quantity = presentQuantity(item);
            String unitPrice = String.format("%.2f", item.getPrice());
            line += "  " + unitPrice + " * " + quantity + "\n";
        }
        return line;
    }

    private static String presentQuantity(ReceiptItem item) {
        return ProductUnit.Each.equals(item.getProduct().getUnit())
                ? String.format("%x", (int) item.getQuantity())
                : String.format("%.3f", item.getQuantity());
    }

    private static String getWhitespace(int whitespaceSize) {
        StringBuilder whitespace = new StringBuilder();
        for (int i = 0; i < whitespaceSize; i++) {
            whitespace.append(" ");
        }
        return whitespace.toString();
    }
}
