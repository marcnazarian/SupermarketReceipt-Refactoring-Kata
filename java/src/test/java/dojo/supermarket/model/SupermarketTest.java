package dojo.supermarket.model;

import dojo.supermarket.ReceiptPrinter;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class SupermarketTest {

    private SupermarketCatalog catalog;
    private Product toothbrush;
    private Product apples;
    private ShoppingCart cart;
    private Teller teller;

    @BeforeEach
    private void setUp() {
        catalog = new FakeCatalog();
        toothbrush = new Product("toothbrush", ProductUnit.Each);
        catalog.addProduct(toothbrush, 0.99);
        apples = new Product("apples", ProductUnit.Kilo);
        catalog.addProduct(apples, 1.99);
        cart = new ShoppingCart();
        teller = new Teller(catalog);
    }

    @Test
    void no_special_offer() {
        cart.addItemQuantity(apples, 2.5);
        teller.addSpecialOffer(SpecialOfferType.TenPercentDiscount, toothbrush, 10.0);

        Receipt receipt = teller.checksOutArticlesFrom(cart);

        Approvals.verify(new ReceiptPrinter().printReceipt(receipt));
    }

    @Test
    void buy_two_toothbrushes_get_one_free_normal_toothbrush_price_is_0_99() {
        cart.addItemQuantity(toothbrush, 2);
        teller.addSpecialOffer(SpecialOfferType.TwoForAmount, toothbrush, 0.99);

        Receipt receipt = teller.checksOutArticlesFrom(cart);

        Approvals.verify(new ReceiptPrinter().printReceipt(receipt));
    }

    @Test
    void discount_20_percent_on_apples_normal_price_is_1_99_per_kilo() {
        cart.addItemQuantity(apples, 2.5);
        teller.addSpecialOffer(SpecialOfferType.TenPercentDiscount, apples, 20.0);

        Receipt receipt = teller.checksOutArticlesFrom(cart);

        Approvals.verify(new ReceiptPrinter().printReceipt(receipt));
    }


//            10% discount on rice, normal price €2.49 per bag
//    Five tubes of toothpaste for €7.49, normal price €1.79
//    Two boxes of cherry tomatoes for €0.99, normal price €0.69 per box.
}
