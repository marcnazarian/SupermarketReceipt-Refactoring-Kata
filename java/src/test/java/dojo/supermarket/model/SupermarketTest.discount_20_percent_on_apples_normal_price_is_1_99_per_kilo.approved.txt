apples                              4.98
  1.99 * 2.500
20.0% off(apples)                  -1.00

Total:                              3.98